src/Symfony/Component/AssetMapper/ImportMap/ImportMapEntry.php

```diff
@@ -18,22 +18,75 @@
  */
 final class ImportMapEntry
 {
-    public function __construct(
-        public readonly string $importName,
-        /**
-         * The path to the asset if local or downloaded.
-         */
-        public readonly ?string $path = null,
-        public readonly ?string $version = null,
-        public readonly ImportMapType $type = ImportMapType::JS,
-        public readonly bool $isEntrypoint = false,
-        public readonly ?string $packageName = null,
-        public readonly ?string $filePath = null,
+    public readonly string $importName;
+    public readonly ImportMapType $type;
+    /**
+     * A logical path, relative path or absolute path to the file.
+     */
+    public readonly string $path;
+    public readonly bool $isEntrypoint;
+
+    // remote-only
+    public readonly string $version;
```

Comment:
Should this be a public property in an uninitialized state for remote ones ? This looks like a weird API (and no SA tool out there will tell you to handle that case).

I'd rather keep them as nullable properties.

Reply:
Note that if you want it to be even nicer for SA tools, you could add the following phpdoc on the `isRemotePackage` method (once the property is nullable rather than left uninitialized):

```php
/**
 * @psalm-assert-if-true !null $this->version
 * @psalm-assert-if-true !null $this->packageModuleSpecifier
 */
```

Reply:
I did that weird api on purpose… but it is weird. Previously, we made these values null for local packages. But that made code around the component more delicate: if we made a mistake and tried to reference version for a local package, instead of an error, other more silent problems would happen (since we would get null when and use it in a context expecting a string). 

The other 2 API ideas I had were: (1) create a RemoteImportmapEntry for remote and then type-hint that or check instanceof around the code and (2) convert ImportMapEntry to getters with private properties so we can throw a clear error when accessing version for a local package.

WDYT?

Reply:
Having getters with a clear error is much better than having a TypeError due to uninitialized properties.

However, this would still not allow SA tools to help you write correct code (as code triggering those errors would still not be correct).

Personally, I would either go with the solution with `@psalm-assert-if-true` on `isRemotePackage` (allowing Psalm to know that checking for remote packages ensures we get non-null values) or a separate class having those 2 additional properties (which has the benefit of being usable in types when a function supports only remote entries)